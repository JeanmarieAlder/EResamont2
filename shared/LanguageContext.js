import { createContext } from "react";

export const LanguageContext = createContext();
// 1 is French
// 2 is Italian
// 3 is English
// 4 is Deutsch
