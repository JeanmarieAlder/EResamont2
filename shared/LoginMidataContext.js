import React, { useState, useEffect, useContext } from "react";
import { ToastAndroid } from "react-native";
import * as AppAuth from "expo-app-auth";
import * as SecureStore from "expo-secure-store";
import requestGetMidata from "../utils/requestGetMidata";

export const Auth2Context = React.createContext();
export const useAuth2 = () => useContext(Auth2Context);
export const Auth2Provider = ({ children }) => {
  const StorageKey = "EResamont2MidataOAuthKey";
  const config = {
    issuer: "https://test.midata.coop/fhir/metadata",
    clientId: "eresamont2-test",
    serviceConfiguration: {
      authorizationEndpoint: "https://test.midata.coop/authservice",
      tokenEndpoint: "https://test.midata.coop/v1/token",
      revocationEndpoint: "https://test.midata.coop/v1/token"
    },
    redirectUrl: AppAuth.OAuthRedirect + ":/oauthredirect"
  };

  const [authState, setAuthState] = useState(null);
  const [userName, setUserName] = useState(null);

  useEffect(() => {
    console.log("LoginMidataContext: useEffect()");
    (async () => {
      let cachedAuth = await getCachedAuthAsync();
      if (cachedAuth && !authState) {
        setAuthState(cachedAuth);
      }
    })();
  }, []);

  const signInAsync = async () => {
    console.log("signInAsync()");
    let authState = await AppAuth.authAsync(config);
    await cacheAuthAsync(authState);
    console.log("signInAsync", authState);
    setAuthState(authState);
    return authState;
  };

  const signOutAsync = async ({ accessToken }) => {
    try {
      await AppAuth.revokeAsync(config, {
        token: accessToken,
        isClientIdProvided: true
      });
      await SecureStore.deleteItemAsync(StorageKey);
      console.log("sign out completed");
      ToastAndroid.show("Sign out completed", ToastAndroid.SHORT);
      setAuthState(null);
      return null;
    } catch (e) {
      console.error(`Failed to revoke token: ${e.message}`);
    }
  };

  const cacheAuthAsync = async authState => {
    return await SecureStore.setItemAsync(
      StorageKey,
      JSON.stringify(authState)
    );
  };

  const getCachedAuthAsync = async () => {
    let value = await SecureStore.getItemAsync(StorageKey);
    let authState = JSON.parse(value);
    console.log("getCachedAuthAsync", authState);
    if (authState) {
      if (checkIfTokenExpired(authState)) {
        return refreshAuthAsync(authState);
      } else {
        return authState;
      }
    }
    return null;
  };

  const checkIfTokenExpired = ({ accessTokenExpirationDate }) => {
    return new Date(accessTokenExpirationDate) < new Date();
  };

  const refreshAuthAsync = async ({ refreshToken }) => {
    let authState = await AppAuth.refreshAsync(config, refreshToken);
    console.log("refreshAuth", authState);
    await cacheAuthAsync(authState);
    return authState;
  };

  const getLoggedUserName = async () => {
    if (authState) {
      let response = await requestGetMidata(
        "https://test.midata.coop/fhir/Patient/" +
          authState.additionalParameters.patient,
        authState,
        signInAsync
      );
      if (response) {
        setUserName(response.name[0].given[0]);
        return true;
      }
    }
    return false;
  };

  return (
    <Auth2Context.Provider
      value={{
        authState,
        userName,
        signInAsync,
        signOutAsync,
        cacheAuthAsync,
        getCachedAuthAsync,
        checkIfTokenExpired,
        refreshAuthAsync,
        getLoggedUserName
      }}
    >
      {children}
    </Auth2Context.Provider>
  );
};
