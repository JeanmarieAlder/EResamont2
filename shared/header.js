import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { MaterialIcons } from "@expo/vector-icons";

export default function Header({ title, navigation }) {
  const openMenu = () => {
    navigation.openDrawer();
  };

  return (
    <View style={styles.header}>
      <MaterialIcons
        name="menu"
        size={28}
        onPress={openMenu}
        style={styles.icon}
      />
      <View>
        <Text style={styles.headerText}>{title}</Text>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  header: {
    flexDirection: "row",
    justifyContent: "flex-end",
    flex: 1
  },
  headerText: {
    fontWeight: "bold",
    fontSize: 20,
    color: "#fff",
    alignItems: "flex-end",
    letterSpacing: 1,
    marginLeft: 10
  },
  icon: { color: "#fff" }
});
