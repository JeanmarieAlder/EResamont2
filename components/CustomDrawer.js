import React, { useContext, useState, useEffect } from "react";
import {
  Text,
  View,
  TouchableOpacity,
  ImageBackground,
  StyleSheet,
  AsyncStorage
} from "react-native";
import * as AppAuth from "expo-app-auth";
import { AuthSession } from "expo";
import { globalStyles } from "../styles/global";
import { LanguageContext } from "../shared/LanguageContext";
import { LoadingContext } from "../shared/LoadingContext";
import ButtonView from "./ButtonView";
import storage from "../utils/storage";
import requestPage from "../utils/requestPage";
import { Icon } from "react-native-elements";
import { Alert, ToastAndroid } from "react-native";
import authMidata from "../utils/authMidata";

export default function CustomDrawer({ navigation }) {
  const { language, setLanguage } = useContext(LanguageContext);
  const { loading, setLoading } = useContext(LoadingContext);

  //login state
  let [authState, setAuthState] = useState(null);
  let StorageKey = "@EResamont2:MidataOAuthKey";

  useEffect(() => {
    console.log("CustomDrawer: useEffect()");
    (async () => {
      let cachedAuth = await getCachedAuthAsync();
      if (cachedAuth && !authState) {
        setAuthState(cachedAuth);
      }
    })();
  }, []);

  async function getCachedAuthAsync() {
    let value = await AsyncStorage.getItem(StorageKey);
    let authState = JSON.parse(value);
    console.log("getCachedAuthAsync", authState);
    if (authState) {
      if (checkIfTokenExpired(authState)) {
        return refreshAuthAsync(authState);
      } else {
        return authState;
      }
    }
    return null;
  }

  function checkIfTokenExpired({ accessTokenExpirationDate }) {
    return new Date(accessTokenExpirationDate) < new Date();
  }

  async function refreshAuthAsync({ refreshToken }) {
    let authState = await AppAuth.refreshAsync(config, refreshToken);
    console.log("refreshAuth", authState);
    await cacheAuthAsync(authState);
    return authState;
  }
  let config = {
    issuer: "https://test.midata.coop/fhir/metadata",
    clientId: "eresamont2-test",
    serviceConfiguration: {
      authorizationEndpoint: "https://test.midata.coop/authservice",
      tokenEndpoint: "https://test.midata.coop/v1/token"
    }
  };

  async function signInAsync() {
    console.log("signInAsync()");
    console.log(AppAuth.OAuthRedirect);
    let authState = await AppAuth.authAsync(config);
    await cacheAuthAsync(authState);
    console.log("signInAsync", authState);
    return authState;
  }

  async function signOutAsync({ accessToken }) {
    try {
      await AppAuth.revokeAsync(config, {
        token: accessToken,
        isClientIdProvided: true
      });
      await AsyncStorage.removeItem(StorageKey);
      return null;
    } catch (e) {
      alert(`Failed to revoke token: ${e.message}`);
    }
  }

  async function cacheAuthAsync(authState) {
    return await AsyncStorage.setItem(StorageKey, JSON.stringify(authState));
  }

  const homeClick = page => () => {
    navigation.navigate(page);
    navigation.toggleDrawer();
  };

  const languageClick = languageSelected => () => {
    storage.setLanguageSetting(languageSelected);
    setLanguage(languageSelected);
    navigation.closeDrawer();
  };
  const checkUpdate = async () => {
    let data = await requestPage.fetchUpdatedContent(null);
    console.log(data.length);
    if (data.length === 0) {
      ToastAndroid.show("Data already up to date", ToastAndroid.SHORT);
    } else {
      let res = await storage.updateStoragePages(data);
      res === true && ToastAndroid.show("Data updated", ToastAndroid.SHORT);
      setLoading(true);
    }
  };
  const clearData = async () => {
    await storage.removeAllStoragePages();
    let info = await storage.checkStoragePages();
    if (info === false) {
      ToastAndroid.show("Local data cleared", ToastAndroid.SHORT);
      setLoading(true);
    }
  };
  const confirmClearDataClick = async () => {
    Alert.alert(
      "Confirm local data deletion",
      "Are you sure?",
      [
        { text: "YES", onPress: () => clearData() },
        {
          text: "NO",
          style: "cancel"
        }
      ],
      { cancelable: true }
    );
  };

  const confirmClearScoreClick = async idQuizz => {
    let result = await storage.getQuizScore(idQuizz);
    if (result) {
      Alert.alert(
        "Confirm local score deletion",
        "Current scores are: " + JSON.stringify(result),
        [
          { text: "YES", onPress: () => storage.deleteScoreData(idQuizz) },
          {
            text: "NO",
            style: "cancel"
          }
        ],
        { cancelable: true }
      );
    }
  };
  const checkInternet = async () => {
    let connectionInfo = await requestPage.checkConnection();
    if (connectionInfo && connectionInfo.isInternetReachable) {
      ToastAndroid.show("Test: Online mode", ToastAndroid.SHORT);
    } else {
      ToastAndroid.show("Test: Offline mode", ToastAndroid.SHORT);
    }
  };
  return (
    <ImageBackground
      source={require("../assets/images/mountain.jpg")}
      style={globalStyles.mountainBackgroundImage}
    >
      <View style={globalStyles.drawerContainer}>
        <View style={globalStyles.drawerTop}>
          <TouchableOpacity
            onPress={homeClick("EResamont")}
            testID={"cd-button-home"}
          >
            <Text style={globalStyles.drawerTitle}>E-Res@mont</Text>
          </TouchableOpacity>
          <View style={globalStyles.drawerTopMenu}>
            <TouchableOpacity onPress={confirmClearDataClick}>
              <Text style={globalStyles.drawwerTopMenuText}>
                Clear local storage
              </Text>
              <View style={localStyles.topMenuDivider} />
            </TouchableOpacity>
            <TouchableOpacity onPress={checkUpdate} testID={"cd-button-update"}>
              <Text style={globalStyles.drawwerTopMenuText}>
                Check for update
              </Text>
              <View style={localStyles.topMenuDivider} />
            </TouchableOpacity>
            <TouchableOpacity onPress={() => confirmClearScoreClick(95)}>
              <Text style={globalStyles.drawwerTopMenuText}>
                Clear lake louise data
              </Text>
              <View style={localStyles.topMenuDivider} />
            </TouchableOpacity>
            <TouchableOpacity onPress={() => confirmClearScoreClick(100)}>
              <Text style={globalStyles.drawwerTopMenuText}>
                Clear oxygen data
              </Text>
              <View style={localStyles.topMenuDivider} />
            </TouchableOpacity>
            <TouchableOpacity onPress={() => checkInternet()}>
              <Text style={globalStyles.drawwerTopMenuText}>
                Check internet
              </Text>
              <View style={localStyles.topMenuDivider} />
            </TouchableOpacity>
            <TouchableOpacity
              onPress={async () => {
                const authState = await signInAsync();
                setAuthState(authState);
              }}
            >
              <Text style={globalStyles.drawwerTopMenuText}>Midata Login</Text>
              <View style={localStyles.topMenuDivider} />
            </TouchableOpacity>
            <TouchableOpacity onPress={() => console.log(authState)}>
              <Text style={globalStyles.drawwerTopMenuText}>
                Auth State (debug)
              </Text>
              <View style={localStyles.topMenuDivider} />
            </TouchableOpacity>
          </View>
        </View>
        <View style={globalStyles.drawerButtons}>
          <ButtonView
            value="FranÃ§ais"
            style={globalStyles.drawerLanguageButton}
            onPress={languageClick(1)}
          />
          <ButtonView
            value="Italiano"
            style={globalStyles.drawerLanguageButton}
            onPress={languageClick(2)}
          />
          <ButtonView
            value="English"
            style={globalStyles.drawerLanguageButton}
            onPress={languageClick(3)}
          />
          <ButtonView
            value="Deutsch"
            style={globalStyles.drawerLanguageButton}
            onPress={() => alert("Translation coming soon")}
          />
        </View>
      </View>
    </ImageBackground>
  );
}
const localStyles = StyleSheet.create({
  topMenuDivider: {
    borderBottomColor: "black",
    borderBottomWidth: 1,
    marginTop: 5,
    marginBottom: 20
  }
});

//deleteScoreData
