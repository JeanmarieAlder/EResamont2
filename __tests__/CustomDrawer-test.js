import * as React from "react";
import renderer from "react-test-renderer";
import { LanguageContext } from "../shared/LanguageContext";

import CustomDrawer from "../components/CustomDrawer";

describe("CustomDrawer", () => {
  const language = 1; //French
  it(`renders correctly`, () => {
    const tree = renderer
      .create(
        <LanguageContext.Provider value={{ language }}>
          <CustomDrawer />
        </LanguageContext.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
