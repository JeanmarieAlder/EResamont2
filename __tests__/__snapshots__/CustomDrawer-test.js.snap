// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomDrawer renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
        "padding": 25,
        "paddingTop": 50,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(255,255,255, 0.7)",
          "borderColor": "#232323",
          "borderRadius": 5,
          "borderStyle": "solid",
          "borderWidth": 0.7,
          "flex": 1,
          "padding": 10,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="cd-button-home"
      >
        <Text
          style={
            Object {
              "color": "#232323",
              "fontFamily": "EuphemiaUCAS-Bold",
              "fontSize": 25,
              "fontWeight": "bold",
            }
          }
        >
          E-Res@mont
        </Text>
      </View>
      <View
        style={
          Object {
            "marginTop": 25,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#232323",
                "fontFamily": "EuphemiaUCAS-Bold",
                "fontSize": 16,
                "textAlign": "left",
              }
            }
          >
            Home
          </Text>
          <View
            style={
              Object {
                "borderBottomColor": "black",
                "borderBottomWidth": 1,
                "marginBottom": 10,
                "marginTop": 5,
              }
            }
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#232323",
                "fontFamily": "EuphemiaUCAS-Bold",
                "fontSize": 16,
                "textAlign": "left",
              }
            }
          >
            Midata Settings
          </Text>
          <View
            style={
              Object {
                "borderBottomColor": "black",
                "borderBottomWidth": 1,
                "marginBottom": 10,
                "marginTop": 5,
              }
            }
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#232323",
                "fontFamily": "EuphemiaUCAS-Bold",
                "fontSize": 16,
                "textAlign": "left",
              }
            }
          >
            Clear local storage
          </Text>
          <View
            style={
              Object {
                "borderBottomColor": "black",
                "borderBottomWidth": 1,
                "marginBottom": 10,
                "marginTop": 5,
              }
            }
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="cd-button-update"
        >
          <Text
            style={
              Object {
                "color": "#232323",
                "fontFamily": "EuphemiaUCAS-Bold",
                "fontSize": 16,
                "textAlign": "left",
              }
            }
          >
            Check for update
          </Text>
          <View
            style={
              Object {
                "borderBottomColor": "black",
                "borderBottomWidth": 1,
                "marginBottom": 10,
                "marginTop": 5,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 1,
          "flexDirection": "column-reverse",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#232323",
            "borderRadius": 10,
            "height": 40,
            "justifyContent": "center",
            "marginTop": 10,
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
        testID="button-view-Français"
      >
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "EuphemiaUCAS-Bold",
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        >
          Français
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#232323",
            "borderRadius": 10,
            "height": 40,
            "justifyContent": "center",
            "marginTop": 10,
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
        testID="button-view-Italiano"
      >
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "EuphemiaUCAS-Bold",
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        >
          Italiano
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#232323",
            "borderRadius": 10,
            "height": 40,
            "justifyContent": "center",
            "marginTop": 10,
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
        testID="button-view-English"
      >
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "EuphemiaUCAS-Bold",
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        >
          English
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#232323",
            "borderRadius": 10,
            "height": 40,
            "justifyContent": "center",
            "marginTop": 10,
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
        testID="button-view-Deutsch"
      >
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontFamily": "EuphemiaUCAS-Bold",
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        >
          Deutsch
        </Text>
      </View>
    </View>
  </View>
</View>
`;
